Fix LineProfile bug where wrong curve length was used

mikebind opened the pull request:
LineProfile used the curve length in mm to calculate the sampling distance between resampled points used to probe to get the profile values, however, the curve which is resampled has been converted to use IJK coordinates instead.  Therefore we need to use the curve length in IJK space which calculating the sampling distance, otherwise, we will get the wrong number of sampling points. I discovered this when the probe filter returned 378 sampled points for an input lineResolution of 100 on the image I was working with. The only change in this PR is to calculate the curve length in IJK coordinates and use that instead when calculating the sampling distance.

lassoan commented:
Good catch! The fix requires some more work, as the issue is actually that resampling is done in IJK space (`slicer.vtkMRMLMarkupsCurveNode.ResamplePoints(curvePoints_IJK, sampledCurvePoints_IJK, samplingDistance, closedCurve)`), which does not make sense, because we want equal point distances in physical space (we don't want the physical distance between sample points depend on direction of the line segment).

Could you update the code to resample points in RAS coordinate system and then convert the coordinates to IJK?

mikebind commented:
Got it, that makes more sense.  I misread the comments in the file as indicating that the resampling had to be uniformly spaced in IJK coordinates in order for the VTK filter to work properly.   Also, since I have been dealing only with single line segments, uniform spacing in IJK was still uniform in the image volume RAS space, but you are correct that if the profile was to be taken along multiple line segments with differing orientations (e.g. along a curve rather than a straight line) uniform IJK sampling will not lead to uniform RAS sampling if the image volume is anisotropic.  I will correct the code to resample uniformly in RAS space and then convert to IJK to feed to the probe filter.

mikebind commented:
After seeing some more odd behavior when I was testing my fix, I did more testing, and think can now characterize when vtkProbeFilter returns values which I think are incorrect. I found that you had already discovered that vtkProbeFilter considers the bounding box of the image to only extend to contain all voxel centers, excluding a 1/2 voxel shell on the outer boundary of the actual image bounding box.  As you pointed out, this is a particular problem for 2D images, because that 1/2 voxel shell includes the whole image except for an infinitesimal plane running though the center. Probe points within the shell but which don't manage to cross the plane are treated as if they were outside the image, and always return a probed value of zero. What was not clear to me (and still isn't), was why crossing the central plane led to a different result.  If vtkProbeFilter thought those points were outside the volume, why, even if they were on opposite side of the plane, would they get interpolated probed values? I tried to understand the vtkProbeFilter source code, but C++ remains too opaque for me to follow it well, so I still don't really understand, beyond that we avoid an early return by crossing the bounding box. As long as we cross the bounding box, I reproduce the behavior where correct values are returned for probe points in the outer shell (for 3D as well as 2D images). 
In addition to that desirable behavior, I found that, as long as one probed point is inside that outer shell (so that returning early with all zeros is avoided) vtkProbeFilter also returns probe values which function something like nearest neighbor extrapolation for an additional 1/2 voxel shell OUTSIDE the image volume boundary, which seems to me to be an incorrect behavior. Points truly outside the image volume should return 0 for a probed value. Here is a [gist demonstrating these findings, just run from Slicer python interactor.](https://gist.github.com/mikebind/86a3d8e0b14036a3d35c3277470cb5c4).  Let me know how you think would be best to handle the situation. I see a few options 
* Ignore it, figuring that, for most use cases, a half-voxel shell too much or too little of probed values will be inconsequential
* keep similar to current behavior, where 2D images are handled as a special case, but the borders of 3D images are not, and the behavior of probed values near these borders are left to the inconsistent whims of vtkProbeFilter
* detect probe points near boundaries as special cases for both 2D and 3D images (and ND?) and handle them in a way so that they return correct and consistent results (in which case we would need to decide exactly what those should be)
I'm also open to other suggestions.

lassoan commented:
>  What was not clear to me (and still isn't), was why crossing the central plane led to a different result. 

The root cause of the problem is that vtkProbeFilter can only interpolate, it cannot extarpolate. See detailed explanation here: https://discourse.vtk.org/t/bounding-box-of-vtkimagedata/2178

> Let me know how you think would be best to handle the situation. I see a few options

The current behavior is good for now, keep handling single-slice volumes as a special case. If a volume has two or more slices then there is a generous sized-space between the first and last slices, so even if there is numerical imprecision it is reasonable to require the user to place the points between the centers of the first and last planes. If a volume has only one slice then we cannot ask users to draw the line inside the volume, because there is no inside for a single-slice volume.

For a more robust and consistent behavior, we would need a vtkImageProbeFilter that David Gobbi described in the VTK forum discussion, which would be able to extrapolate from boundary points by a half voxel.

lassoan commented:
vtkProbeFilter actually can only interpolate, but it is very hard to detect if the sampling object is inside the dataset (it just checks if the bounding boxes intersect) so vtkProbeFilter can be tricked into a slight extrapolation. Of course this extrapolation is nothing sophisticated, it is just some weighted average or something like that, so essentially it returns the linear interpolation between closest points in the image.

Indeed, extending the image could be one solution for this, but it would be an overkill (it would be more complex, potentially much, much slower and could use a lot of memory for a big image), and it would still be just a workaround. If you would like a clean solution then you could implement vtkImageProbeFilter.

mikebind commented:
Thanks, I appreciate the extra explanation.  I implemented the suggested approach (maintain status quo: 2D images are handled as a special case and 3D images retain inconsistent behavior on outer shell and near outer boundary), with the bug fix that uniform resampling is done in RAS space before converting to IJK coordinates for probing.  I also added a small note about the boundary behavior into the helpText.

lassoan reviewed (approved):
Thank you, this looks good!
